File: Code/Asteroid_Trajectory_Selection/scripts/greedy/analysis/find_all_paths_greedy.m
NUM_ASTEROIDS = length(asteroid_list);

mat_files = dir(fullfile(getRepoRoot(), 'outputs', 'greedy', '*.mat'));

TOTAL_DV_MAP = ones(NUM_ASTEROIDS, NUM_ASTEROIDS, NUM_ASTEROIDS);

DATA = {'dv_total_mission', 'dv_launch_leg1', 'asteroid 1', 'asteroid 2', 'asteroid 3', 'filename';};
for p = 1:length(mat_files)
    i_mins = -1 * ones(1, 1);
    j_mins = -1 * ones(1, 1);
    k_mins = -1 * ones(1, 1);

    load(fullfile(mat_files(p).folder, mat_files(p).name));
    for n = 1:1
        i_min_current = -1;
        j_min_current = -1;
        k_min_current = -1;
        min_delta_v = Inf;
        
        for i = 1:NUM_ASTEROIDS
            for j = 1:NUM_ASTEROIDS
                for k = 1:NUM_ASTEROIDS
                    if i == j || j == k || k == i
                        continue
                    end
    
                    total_dv = 0;
                    for l = 1:3
                        total_dv = total_dv + asteroid_optimized_data(i, j, k, l).dv_total;
                    end
                    TOTAL_DV_MAP(i, j, k) = total_dv;
    
                    if min_delta_v > total_dv
                        avoid = 0;
                        for w = 1:n
                            if i_mins(w) == i && j_mins(w) == j && k_mins(w) == k
                                avoid = 1;
                                break;
                            end
                        end
    
                        if avoid == 1
                            continue;
                        end
    
                        min_delta_v = total_dv;
                        i_min_current = i;
                        j_min_current = j;
                        k_min_current = k;
                    end
                end
            end
        end
        
        
    end

    i_mins(n) = i_min_current;
        j_mins(n) = j_min_current;
        k_mins(n) = k_min_current;
        % Display the minimum delta_v and corresponding asteroid indices

        DATA(end+1, :) = {min_delta_v, asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).dv_launch,...
            asteroid_list(i_mins(n)).NAME, asteroid_list(j_mins(n)).NAME, asteroid_list(k_mins(n)).NAME, mat_files(p).name};
        fprintf('FILE: %s | Total mission dv (3 legs): %.2f km/s (leg 1 Earth launch dv %.2f) | %s (INDEX %d) | %s (INDEX %d) | %s (INDEX %d)\n', ...
            mat_files(p).name, min_delta_v, asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).dv_launch, asteroid_list(i_mins(n)).NAME, i_mins(n),...
            asteroid_list(j_mins(n)).NAME, j_mins(n),...
            asteroid_list(k_mins(n)).NAME, k_mins(n));
        fprintf(['LAUNCH AT %s (dv %.2f) | FLYBY %s AT %s (dv %.2f) | ARRIVE AT %s (dv %.2f)\n' ...
            'LEAVE AT %s (dv %.2f) | FLYBY %s AT %s (dv %.2f) | ARRIVE AT %s (dv %.2f) \n' ...
            'LEAVE AT %s (dv %.2f) | FLYBY %s AT %s (dv %.2f) | ARRIVE AT %s (dv %.2f) \n\n\n'], ...
            cspice_et2utc(norm(asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).et_launch), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).dv_launch, ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).FLYBY_BODY, ...
            cspice_et2utc(norm(asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).et_flyby), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).dv_arrive, ...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).et_goal), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 1).dv_goal, ...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).et_launch), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).dv_launch, ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).FLYBY_BODY, ...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).et_flyby), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).dv_arrive,...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).et_goal), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 2).dv_goal,...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).et_launch), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).dv_launch,...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).FLYBY_BODY, ...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).et_flyby), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).dv_arrive,...
            cspice_et2utc((asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).et_goal), 'C', 1), ...
            asteroid_optimized_data(i_mins(n), j_mins(n), k_mins(n), 3).dv_goal);
end

outCsvDir = fullfile(getRepoRoot(), 'outputs', 'tables');
if ~isfolder(outCsvDir)
    mkdir(outCsvDir);
end
writecell(DATA, fullfile(outCsvDir, 'asteroid_optimized_data_table.csv'));

File: Python/asteroid_trajectory_selection/greedy/find_all_paths_greedy.py
"""Load all greedy optimized data files and find the best path in each."""

import glob
import os
import pickle

import numpy as np
import pandas as pd
import spiceypy

from ..load_kernels import load_kernels


def main():
    """Main entry point."""
    asteroid_list = load_kernels("NOTABLE_ASTEROID_BSPs", "generic_kernels")
    NUM_ASTEROIDS = len(asteroid_list)

    pkl_files = sorted(glob.glob(os.path.join("greedy_asteroid_paths", "*.pkl")))

    TOTAL_DV_MAP = np.ones((NUM_ASTEROIDS, NUM_ASTEROIDS, NUM_ASTEROIDS))

    DATA = []

    for p, pkl_path in enumerate(pkl_files):
        filename = os.path.basename(pkl_path)

        with open(pkl_path, 'rb') as f:
            asteroid_optimized_data = pickle.load(f)

        i_mins = [-1]
        j_mins = [-1]
        k_mins = [-1]

        for n in range(1):  # MATLAB 1:1 = single iteration
            i_min_current = -1
            j_min_current = -1
            k_min_current = -1
            min_delta_v = np.inf

            for i in range(NUM_ASTEROIDS):
                for j in range(NUM_ASTEROIDS):
                    for k in range(NUM_ASTEROIDS):
                        if i == j or j == k or k == i:
                            continue

                        if (i, j, k) not in asteroid_optimized_data:
                            continue

                        data = asteroid_optimized_data[(i, j, k)]

                        # total_dv = leg1.dv_goal + leg2.dv_total + leg3.dv_total
                        total_dv = data[0]['dv_goal']
                        for l in range(1, 3):
                            total_dv += data[l]['dv_total']

                        TOTAL_DV_MAP[i, j, k] = total_dv

                        if min_delta_v > total_dv:
                            avoid = False
                            for w in range(n):
                                if (i_mins[w] == i and j_mins[w] == j
                                        and k_mins[w] == k):
                                    avoid = True
                                    break
                            if avoid:
                                continue

                            min_delta_v = total_dv
                            i_min_current = i
                            j_min_current = j
                            k_min_current = k

        i_mins[0] = i_min_current
        j_mins[0] = j_min_current
        k_mins[0] = k_min_current

        if i_min_current < 0 or j_min_current < 0 or k_min_current < 0:
            continue

        data = asteroid_optimized_data[(i_mins[0], j_mins[0], k_mins[0])]

        dv_launch = data[0]['dv_launch']

        DATA.append({
            'dv_asteroids': min_delta_v,
            'dv_launch': dv_launch,
            'asteroid 1': asteroid_list[i_mins[0]]['NAME'],
            'asteroid 2': asteroid_list[j_mins[0]]['NAME'],
            'asteroid 3': asteroid_list[k_mins[0]]['NAME'],
            'filename': filename,
        })

        # Print detailed info
        print(f'FILE: {filename} | ASTEROID TRANSFERS dv: {min_delta_v:.2f} km/s '
              f'(EARTH LAUNCH DV {dv_launch:.2f}) | '
              f'{asteroid_list[i_mins[0]]["NAME"]} (INDEX {i_mins[0]}) | '
              f'{asteroid_list[j_mins[0]]["NAME"]} (INDEX {j_mins[0]}) | '
              f'{asteroid_list[k_mins[0]]["NAME"]} (INDEX {k_mins[0]})')

        for leg_idx in range(3):
            leg = data[leg_idx]
            launch_utc = spiceypy.et2utc(
                abs(leg['et_launch']), 'C', 1)
            flyby_utc = spiceypy.et2utc(
                abs(leg['et_flyby']), 'C', 1)
            goal_utc = spiceypy.et2utc(leg['et_goal'], 'C', 1)

            print(f'  LEG {leg_idx + 1}: '
                  f'LAUNCH AT {launch_utc} (dv {leg["dv_launch"]:.2f}) | '
                  f'FLYBY {leg["FLYBY_BODY"]} AT {flyby_utc} (dv {leg["dv_arrive"]:.2f}) | '
                  f'ARRIVE AT {goal_utc} (dv {leg["dv_goal"]:.2f})')
        print()

    # Write CSV
    df = pd.DataFrame(DATA)
    df.to_csv('asteroid_optimized_data_table.csv', index=False)
    print(f'Wrote asteroid_optimized_data_table.csv ({len(DATA)} rows)')


if __name__ == '__main__':
    main()
